### PR TITLE
[17.0][MIG] sale_blanket_order

### PR DESCRIPTION
**Pull request by niboo-jva**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Thanks for the contribution.

Please preserve commit history following technical method explained in https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0.

If the jump is between several versions, you have to modify the source branch in the main command to accommodate it to this circumstance.

**Review by rousseldenis — APPROVED** [review]
Code review

**Review by Deriman-Alonso — CHANGES_REQUESTED** [review]
Hi, on tree views invisible must be replaced by column_invisible, Thanks

File: sale_blanket_order/models/blanket_orders.py [hunk 1]
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
         return product[field_name] * uom_factor * cur_factor, currency_id.id
 
     def _get_display_price(self, product):
-        # Copied and adapted from the sale module
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py]: Please leave this comment in for traceability purposes.

File: sale_blanket_order/models/blanket_orders.py [hunk 2]
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
...
-            rule_id,
-            self.original_uom_qty,
-            self.product_uom,
-            pricelist.id,
+
+        pricelist_price = self.pricelist_item_id._compute_price(
+            product=self.product_id,
+            quantity=self.original_uom_qty or 1.0,
+            uom=self.product_uom,
+            date=self.order_id.validity_date,
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:565]: I don't think you should use `validity_date` here. The `validity_date` defines the date until which the blanket order is valid. The date entered in the `_compute_price()` function should be the 'date to use for price computation and currency conversions'. I see `fields.Date.today()` is used in earlier versions to compute the price.

File: sale_blanket_order/models/blanket_orders.py [hunk 3]
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
...
+            return pricelist_price
+
+        base_price = self._get_pricelist_price_before_discount()
+
         # negative discounts (= surcharge) are included in the display price
-        return max(base_price, final_price)
+        return max(base_price, pricelist_price)
+
+    def _get_pricelist_price_before_discount(self):
+        """Compute the price used as base for the pricelist price computation.
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:582]: This is a copy & paste from the sale module. Mentioning it would be helpful for future migrations.

E.g.: `# Copied and adapted from the sale module`

File: sale_blanket_order/models/blanket_orders.py [hunk 4]
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
...
+        :rtype: float
+        """
+        self.ensure_one()
+        self.product_id.ensure_one()
+
+        return self.pricelist_item_id._compute_price_before_discount(
+            product=self.product_id,
+            quantity=self.product_uom_qty or 1.0,
+            uom=self.product_uom,
+            date=self.order_id.date_order,
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:594]: `order_id` has no field `date_order`. Copy & paste mistake?

File: sale_blanket_order/models/blanket_orders.py [hunk 5]
```diff
@@ -564,34 +556,44 @@ def _get_real_price_currency(self, product, rule_id, qty, uom, pricelist_id):
         return product[field_name] * uom_factor * cur_factor, currency_id.id
 
     def _get_display_price(self, product):
-        # Copied and adapted from the sale module
         self.ensure_one()
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:559]: Why not include the `self.product_id.ensure_one()` here as well, similar to the function in the sale module?

File: sale_blanket_order/models/blanket_orders.py [hunk 6]
```diff
@@ -656,6 +658,23 @@ def _compute_quantities(self):
                 line.remaining_uom_qty, line.product_id.uom_id
             )
 
+    @api.depends("product_id", "product_uom", "original_uom_qty")
+    def _compute_pricelist_item_id(self):
+        for line in self:
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:663]: Please add `# Copied and adapted from the sale module`

File: sale_blanket_order/models/blanket_orders.py [hunk 7]
```diff
@@ -656,6 +658,23 @@ def _compute_quantities(self):
...
+                or line.display_type
+                or not line.order_id.pricelist_id
+            ):
+                line.pricelist_item_id = False
+            else:
+                line.pricelist_item_id = line.order_id.pricelist_id._get_product_rule(
+                    line.product_id,
+                    quantity=line.original_uom_qty or 1.0,
+                    uom=line.product_uom,
+                    date=line.order_id.validity_date,
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:675]: `validity_date` shouldn't be used here I think (see my other comment). Perhaps best to use `fields.Date.today()`.

File: sale_blanket_order/models/sale_orders.py [hunk 8]
```diff
@@ -61,14 +61,14 @@ def _get_assigned_bo_line(self, bo_lines):
         assigned_bo_line = False
         date_planned = date.today()
         date_delta = timedelta(days=365)
-        for line in bo_lines.filtered(lambda l: l.date_schedule):
+        for line in bo_lines.filtered(lambda bo_line: bo_line.date_schedule):
```
Comment (chrisb-c01) [inline sale_blanket_order/models/sale_orders.py:64]: Is this change really needed?

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 9]
```diff
@@ -37,25 +37,25 @@
                         string="Create Sale Order"
                         type="action"
                         class="btn-primary"
-                        attrs="{'invisible': [('state', '!=', 'open')]}"
+                        invisible="state != 'open'"
                     />
                     <button
                         name="set_to_draft"
-                        states="expired,done"
+                        invisible="state != ['expired', 'done']"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:44]: Do you mean `state not in ['expired', 'done']`?

```
>>> state = 'ready'
>>> state != ['expired', 'done']
True
>>> state not in ['expired', 'done']
True
>>> state = 'done'
>>> state != ['expired', 'done']
True
>>> state not in ['expired', 'done']
False
```

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 10]
```diff
@@ -146,10 +143,8 @@
                                     <field name="sequence" widget="handle" />
                                     <field
                                         name="product_id"
-                                        attrs="{
-                                            'required': [('display_type', '=', False)],
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:147]: Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 11]
```diff
@@ -181,10 +176,8 @@
                                             'quantity': original_uom_qty,
                                             'company_id': parent.company_id
                                         }"
-                                        attrs="{
-                                            'required': [('display_type', '=', False)],
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:180]: Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 12]
```diff
@@ -198,10 +191,8 @@
                                     />
                                     <field
                                         name="price_unit"
-                                        attrs="{
-                                            'required': [('display_type', '=', False)],
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:195]: Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 13]
```diff
@@ -214,17 +205,13 @@
                                         domain="[('type_tax_use','=','sale')]"
                                         context="{'default_type_tax_use': 'sale'}"
                                         options="{'no_create': True}"
-                                        attrs="{
-                                            'required': [('display_type', '=', False)],
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        required="display_type == False"
+                                        invisible="display_type == True"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:209]: Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 14]
```diff
@@ -214,17 +205,13 @@
...
                                     />
                                     <field
                                         name="price_subtotal"
                                         widget="monetary"
-                                        attrs="{
-                                            'invisible': [('display_type', '=', True)],
-                                        }"
+                                        invisible="display_type == True"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:214]: Please use `column_invisible` see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

File: sale_blanket_order/views/sale_blanket_order_views.xml [hunk 15]
```diff
@@ -284,7 +271,7 @@
                                     <t groups="base.group_multi_company">
                                         <field
                                             name="analytic_account_id"
-                                            attrs="{'readonly': [('sale_count','!=',0),('state','!=','draft')]}"
+                                            readonly="sale_count != 0 or state != 'draft'"
```
Comment (chrisb-c01) [inline sale_blanket_order/views/sale_blanket_order_views.xml:274]: Shouldn't the `or` be and `and` as no or-operator (`'|'`) is present in de `attrs` domain?

File: sale_blanket_order/models/blanket_orders.py [hunk 16]
```diff
@@ -483,6 +472,9 @@ def _compute_amount(self):
         default=False,
         help="Technical field for UX purpose.",
     )
+    pricelist_item_id = fields.Many2one(
+        comodel_name="product.pricelist.item", compute="_compute_pricelist_item_id"
+    )
 
     def name_get(self):
```
Comment (chrisb-c01) [inline sale_blanket_order/models/blanket_orders.py:479]: This should be updated as per https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0 

> If you are overriding name_get method, you should now override _compute_display_name instead, adding the possible new fields in the depends of the method. See https://github.com/odoo/odoo/pull/122085 for more details.

**Review by chrisb-c01 — CHANGES_REQUESTED** [review]

**Comment by chrisb-c01** [comment]
Hi @jvandri thanks for your effort on getting this to v17. Will you have some time to apply the fixes so that we can get this migrated?

File: sale_blanket_order/models/sale_orders.py (hunk 8 above)
Comment (chrisb-c01) [reply sale_blanket_order/models/sale_orders.py:64]: > Is this change really needed?

Apparently this is needed because of a linting error:

> E741 Ambiguous variable name: `l`

File: sale_blanket_order/i18n/it.po [hunk 17]
```diff
@@ -6,7 +6,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Odoo Server 16.0\n"
 "Report-Msgid-Bugs-To: \n"
-"PO-Revision-Date: 2024-03-15 16:34+0000\n"
+"PO-Revision-Date: 2024-01-25 15:34+0000\n"
```
Comment (chrisb-c01) [inline sale_blanket_order/i18n/it.po:9]: Was this intentional (to go back in time)?

**Review by chrisb-c01 — CHANGES_REQUESTED** [review]

**Comment by chrisb-c01** [comment]
I've created a new PR based on this one and applied my changes. https://github.com/OCA/sale-workflow/pull/3255

**Comment by rousseldenis** [comment]
@jvandri We switch to #3255
